### PR TITLE
[Drop-In] Remove `@Experimental` annotation from drop in related files

### PR DESCRIPTION
**Pull request by abhishek1508**
### Description

The implementation covers 
- removing `@ExperimentalPreviewMapboxNavigationAPI` annotation from all `internal` and `public` files related to `libnavui-dropin` module.
- removing `@ExperimentalPreviewMapboxNavigationAPI` from all new components that are exposed outside `libnavui-dropin` module. For ex: `RouteLineComponent`, `RouteLineComponentContract`, `RecenterButtonComponent` etc. It shouldn't be an issue removing the experimental annotation, since these components and their respective contracts are hosted inside `internal` package.
- adding`@ExperimentalPreviewMapboxNavigationAPI` to `MapboxExtendableButton.updateStyles` and it's usage across the SDK. 
- _**not removing**_ `@ExperimentalPreviewMapboxNavigationAPI` from any files related to `ComponentInstaller` and it's usage. `ComponentInstaller` could be subject to modifications and we need to play more with them before declaring them stable.

### CODE REVIEW

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonBinder.kt [hunk 1]
```diff
@@ -33,7 +33,7 @@ import com.mapbox.navigation.ui.voice.view.MapboxAudioGuidanceButton
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.StateFlow
 
-@ExperimentalPreviewMapboxNavigationAPI
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
```
[the remark this answers is not in the document]
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonBinder.kt:36]: This uses `MapboxExtendableButton` which is still `@Experimental`. 

> not removing @ExperimentalPreviewMapboxNavigationAPI from MapboxExtendableButton and it's usage across the SDK. This is because in some cases it's still not possible to render circular buttons with this widget. Issue was reported https://github.com/mapbox/mapbox-navigation-android/issues/6248

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/extensions/MapboxLifecycleExtensions.kt [hunk 2]
```diff
@@ -4,12 +4,10 @@ import androidx.lifecycle.DefaultLifecycleObserver
 import androidx.lifecycle.Lifecycle
 import androidx.lifecycle.LifecycleEventObserver
 import androidx.lifecycle.LifecycleOwner
-import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
 import com.mapbox.navigation.core.MapboxNavigation
 import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
 import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
 
-@ExperimentalPreviewMapboxNavigationAPI
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/extensions/MapboxLifecycleExtensions.kt:12]: You did not update core's current.txt.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt [hunk 3]
```diff
@@ -60,25 +59,24 @@ import com.mapbox.navigation.ui.utils.internal.lifecycle.ViewLifecycleRegistry
...
- * If you prefer the lifecycle of the [ViewModel]s internal to [NavigationView] to be tied to the hosting [Fragment] instead,
- * provide that [Fragment] as an argument. Providing a [Fragment]'s store owner allows to achieve tighter memory control if the resources
- * used by the [NavigationView] shouldn't outlive its hosting [Fragment] (in anticipation of potential [Fragment] recreation).
+ * If you prefer the lifecycle of the [ViewModel]s internal to [NavigationView] to be tied to the
+ * hosting [Fragment] instead, provide that [Fragment] as an argument. Providing a [Fragment]'s
+ * store owner allows to achieve tighter memory control if the resources used by the
+ * [NavigationView] shouldn't outlive its hosting [Fragment] (in anticipation of potential
+ * [Fragment] recreation).
  */
-@ExperimentalPreviewMapboxNavigationAPI
```
Comment (dzinad) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:81]: First https://github.com/mapbox/mapbox-navigation-android/pull/6433/ should be merged or I'll break API.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/MapBinder.kt [hunk 4]
```diff
@@ -38,7 +37,6 @@ import kotlinx.coroutines.FlowPreview
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.combine
 
-@ExperimentalPreviewMapboxNavigationAPI
```
Comment (dzinad) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/MapBinder.kt:41]: Merge conflicts with https://github.com/mapbox/mapbox-navigation-android/pull/6433 are coming...

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/ComponentInstaller.kt [hunk 5]
```diff
@@ -156,6 +155,30 @@ fun ComponentInstaller.recenterButton(
...
 
+/**
+ * Install component that toggles [NavigationCamera] between [NavigationCameraState.OVERVIEW] and
+ * [NavigationCameraState.FOLLOWING] on [button] click.
+ *
+ * The installed component will first use the [NavigationCamera] instance from [CameraModeConfig].
+ * If not available, it will attempt to use the [NavigationCamera] instance managed by the NavigationCamera component.
+ * (see [ComponentInstaller.navigationCamera])
+ */
+@ExperimentalPreviewMapboxNavigationAPI
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/ComponentInstaller.kt:166]: Do you need experimental here and everywhere else in this class?

File: libnavui-speedlimit/src/main/java/com/mapbox/navigation/ui/speedlimit/ComponentInstaller.kt [hunk 6]
```diff
@@ -32,6 +31,7 @@ fun ComponentInstaller.speedLimit(
 /**
  * Speed limit view component configuration class.
  */
+@ExperimentalPreviewMapboxNavigationAPI
```
Comment (dzinad) [inline libnavui-speedlimit/src/main/java/com/mapbox/navigation/ui/speedlimit/ComponentInstaller.kt:34]: The PR is about removing the annotation, should ie be added here?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt (hunk 3 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:81]: Yeah this PR is not gonna merge until next week

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/ComponentInstaller.kt (hunk 5 above)
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/ComponentInstaller.kt:166]: Yes. Refers
> not removing @ExperimentalPreviewMapboxNavigationAPI from any files related to ComponentInstaller and it's usage. ComponentInstaller could be subject to modifications and we need to play more with them before declaring them stable.

File: libnavui-speedlimit/src/main/java/com/mapbox/navigation/ui/speedlimit/ComponentInstaller.kt (hunk 6 above)
Comment (abhishek1508) [reply libnavui-speedlimit/src/main/java/com/mapbox/navigation/ui/speedlimit/ComponentInstaller.kt:34]: Yes. Refers
> not removing @ExperimentalPreviewMapboxNavigationAPI from any files related to ComponentInstaller and it's usage. ComponentInstaller could be subject to modifications and we need to play more with them before declaring them stable.

File: libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt [hunk 7]
```diff
@@ -21,8 +21,8 @@ import com.mapbox.navigation.ui.utils.internal.extensions.measureTextWidth
 /**
  * Button with an icon and expand behaviour.
  */
-@ExperimentalPreviewMapboxNavigationAPI
 @UiThread
+@ExperimentalPreviewMapboxNavigationAPI
```
Comment (tomaszrybakiewicz) [inline libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt:25]: Instead of marking this class experimental, why not just mark`fun updateStyle(@StyleRes style: Int)` `@Deprecated`?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/extensions/NavigationViewContextEx.kt [hunk 8]
```diff
@@ -36,11 +35,6 @@ import com.mapbox.navigation.dropin.tripprogress.TripProgressBinder
 import com.mapbox.navigation.dropin.tripsession.TripSessionComponent
 import kotlinx.coroutines.flow.combine
 
-//
-// Components
```
Comment (tomaszrybakiewicz) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/extensions/NavigationViewContextEx.kt:40]: @abhishek1508 You don't like my comments?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/marker/MapMarkerFactory.kt [hunk 9]
```diff
@@ -10,9 +10,6 @@ import com.mapbox.maps.extension.style.layers.properties.generated.IconAnchor
 import com.mapbox.maps.plugin.annotation.generated.PointAnnotationOptions
 import com.mapbox.navigation.dropin.util.BitmapMemoryCache
 
-/**
- * Factory class for creating all drop-in UI map point annotations.
- */
```
Comment (tomaszrybakiewicz) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/marker/MapMarkerFactory.kt:15]: @abhishek1508 Why did you remove this comment?

**Comment by tomaszrybakiewicz** [comment]
@abhishek1508 The changes that remove `@Experimental` annotation are great. 
However, I have a couple of questions about some other changes:
- Why are we moving and modifying `ComponentInstaller` classes? I think these changes are outside the scope of this PR and should be done separately.
- Why are we removing class description comments?

File: libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt (hunk 7 above)
Comment (abhishek1508) [reply libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt:25]: It sounds a bit funny to release a product in GA with `@deprecated` API. I have rather removed the `Experimental` annotation from the class and only marked that particular API as `Experimental`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/marker/MapMarkerFactory.kt (hunk 9 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/marker/MapMarkerFactory.kt:15]: In the beginning all our classes were public and hence the need for documentation for all. Since we changed the relevant ones to internal, we did not remove the comments. Our code base has not become quite inconsistent as to some `internal` classes have descriptions vs the others don't. This was an attempt to bring consistency. 

However I have put them back for the time being.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/ComponentInstaller.kt [hunk 10]
```diff
@@ -1,4 +1,4 @@
-package com.mapbox.navigation.ui.maps.installer
```
[the remark this answers is not in the document]
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/ComponentInstaller.kt:1]: `ComponentInstaller` files were overlooked on my part when restructuring the package for Drop-In UI. Some `ComponentInstaller` live directly under the root vs the others live under `root/installer`. This was difficult to find as each module was hosting its `ComponentInstaller` in a different place.

**Comment by abhishek1508** [comment]
> @abhishek1508 The changes that remove `@Experimental` annotation are great. However, I have a couple of questions about some other changes:
> 
> * Why are we moving and modifying `ComponentInstaller` classes? I think these changes are outside the scope of this PR and should be done separately.

Answered [here](https://github.com/mapbox/mapbox-navigation-android/pull/6471#discussion_r1018279621)

> * Why are we removing class description comments?

Answered [here](https://github.com/mapbox/mapbox-navigation-android/pull/6471#discussion_r1018276736)

@tomaszrybakiewicz

File: libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt (hunk 7 above)
Comment (tomaszrybakiewicz) [reply libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt]: Or... since the class is still experimental, and to make it stable we need to remove `updateStyle` method,  just delete the method.

Comment (abhishek1508) [reply libnavui-base/src/main/java/com/mapbox/navigation/ui/base/view/MapboxExtendableButton.kt]: Done.

**Comment by abhishek1508** [comment]
@tomaszrybakiewicz All suggestions have been applied. PTAL and let me know if any other changes are needed.

**Review by tomaszrybakiewicz — APPROVED** [review]
LGTM

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/ForwardMapboxNavigation.kt [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ * makes your class appear as if it implements [MapboxNavigationObserver] without exposing the
+ * functions.
+ *
+ * ``` kotlin
+ * val navigationObserver = mapboxNavigationForward(this::onAttached, this::onDetached)
+ * private fun onAttached(mapboxNavigation: MapboxNavigation)
+ * private fun onDetached(mapboxNavigation: MapboxNavigation)
+ * ```
+ */
+fun forwardMapboxNavigation(
```
Comment (Zayankovsky) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/ForwardMapboxNavigation.kt:18]: It's weird that we now have `mapboxNavigationForward` and `forwardMapboxNavigation` with identical implementations. cc @kmadsen

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/ForwardMapboxNavigation.kt:18]: Yeah, I had some direct messages with `abhishek1508` about this. I really like the `mapboxNavigationForward` function in android auto. It is a little similar to `requireMapboxNavigation`

So the plan will be to delete the `mapboxNavigationForward` function in androidauto, in favor of `forwardMapboxNavigation` in the core-sdk. Sorry it's confusing to have them renamed, consequence of iteration.

But this is all intentional
